### DISCUSSION TITLE
How to manage relatedConnection for indirect relations with the prisma plugin?

### DISCUSSION THREAD

**Discussion by abathellier** [hayes/pothos#631] [Q&A] [answered]
Is it possible to use [nestedSelection](https://pothos-graphql.dev/docs/plugins/prisma#indirect-relations-eg-join-tables) for the `relatedConnection` method? If not what is the best way to manage indirect relations for a relay connection with prisma?

Currently i'm doing this (it's inside a `Collection` ObjectType):

```ts
  posts: t.prismaConnection({
    type: "Post",
    cursor: "id",
    resolve: async (query, parent, args, ctx) => {
      return ctx.db.post.findMany({
        ...query,
        where: { collections: { some: { post: { id: parent.id } } } },
      });
    },
  }),
```

Maybe there could be a more efficient way to do this with something like:

```ts
  posts: t.relatedConnection("posts", {
    cursor: "id",
    select: (args, ctx, nestedSelection) => ({
      posts: {
        select: {
          // This will look at what fields are queried on Media
          // and automatically select uploadedBy if that relation is requested
          post: nestedSelection(
            // This arument is the default query for the media relation
            // It could be something like: `{ select: { id: true } }` instead
            true
          ),
        },
      },
    }),
    resolve: (collection) => collection.posts.map(({ post }) => post),
  }),
```

**Comment by hayes** [comment]
working on a solution for this in #671, sorry I didn't respond here sooner. Will follow up with how to use this once its been finished up, there is still some work to do before its ready to go

**Comment by hayes** [comment, marked as answer]
Added some new helpers for this.  Docs are available here: https://pothos-graphql.dev/docs/plugins/prisma#indirect-relations-as-connections

**Reply by toxsick** [reply]
Hey @hayes,

thank you very much for this example! This is exactly what I try to implement right now. I am not sure paging works as expected right now though.

In your example you define `cursor: 'id'` in `mediaConnectionHelpers`. As I understand it, this defines the field in the connection target that is used for paging. This information will be used in the `...mediaConnectionHelpers.getQuery(args, ctx)` call to generate a cursor ( something like `cursor: { id: <mediaId> }` in select portion of the t.connection definition).

This will not result in an error when used with `after: "<some-valid-cursor"`, BUT it will not page correctly, because what that actually does is using the id from the Media model and use that id as a cursor for the PostMedia (Join Table).

Instead the cursor should resolve the id of PostMedia or something like `cursor: { postId: <postId>, mediaId: <mediaId> }` to page correctly I think.

**TLDR Post -> id is used as cursor for paging through PostMedia -> id (other id)**

I hope you understand what I mean or am I missing something here?

Thank you  so much for pothos and regards

Hannes

**Reply by hayes** [reply]
Yup, you are right, this is definitely not correct... I'll look at how to make this work later this weekend

**Reply by toxsick** [reply]
Nice, thank you very much!

A little addition to this. In my use case the join table does not have a primary key `id`, but instead a combined primary key (`@@id([postId, mediaId])` and `@@unique([postId, mediaId] )` in this example)

Paging would only work with a combined cursor in this case (something like `cursor: { postId: <postId>, mediaId: <mediaId> }`)

If this would be supported that would be awesome. Otherwise I have to add an additional `id` column as primary key to the join table for paging.

Thanks a lot and have a nice weekend

**Reply by hayes** [reply]
<img width="868" alt="Screen Shot 2022-11-24 at 3 52 27 PM" src="https://user-images.githubusercontent.com/816527/203875775-7cb2ac02-479b-47d7-82f1-c679212df048.png">

I need to do some testing and cleanup, but I think something like this should work.  It requires a breaking change, but since the current version is actually completely broken I will probably release it as a minor.

**Reply by hayes** [reply]
If you want to test this out, I published a snapshot preview here: https://github.com/hayes/pothos/pull/683

Still need to update docs and test some more edge cases, but I think this is close to a working solution

**Reply by hayes** [reply]
Basically, the connection helper should now be created with the "join table", and include a "select" option that selects the field used as the node, and a resolveNode option to resolve the node.  The helpers for the actual connection field now work slightly differently, but should be simpler overall.  This also makes it easy to add additional fields for your edge objects

**Reply by hayes** [reply]
published a new version that has this refactor.  Hope that works for you!  thanks for checking this out and pointing out the issue so quickly!

**Reply by toxsick** [reply]
Hey @hayes, thank you so much for this super fast fix. From my initial tests it works like a charm and also the code seems much simpler!

**Reply by toxsick** [reply]
I have one remaining question though, how can I get a `totalCount` in this construct?

**Reply by hayes** [reply]
<img width="929" alt="Screen Shot 2022-11-28 at 8 01 30 AM" src="https://user-images.githubusercontent.com/816527/204324706-78770303-2403-4719-9e01-3586132912a6.png">

**Reply by toxsick** [reply]
Hey @hayes, that also worked perfectly. Thanks for that!

Last thing I can't get wrap my head around in this construct is how to get sorting (`orderBy`) working. Everything I add to the `args` gets ignored and the resulting SQL queries do not have an ORDER BY clause.

THANKS and have a nice weekend!

**Reply by hayes** [reply]
something like this should work:

<img width="857" alt="Screen Shot 2023-01-27 at 3 35 02 PM" src="https://user-images.githubusercontent.com/816527/215225361-7b91b689-a3dc-4109-8857-6fa92cf91cfe.png">

**Reply by hayes** [reply]
or with args: 
<img width="755" alt="Screen Shot 2023-01-27 at 3 38 23 PM" src="https://user-images.githubusercontent.com/816527/215225699-a6576640-1d65-4201-86e5-18ee1b77a824.png">

**Reply by hayes** [reply]
obviously this isn't a great pattern, you would likely want an SortDirection enum or something similar, and maybe a SortBy enum for the different values you can sort on, but hopefully this is enough to get you going

**Reply by toxsick** [reply]
AHHH there it belongs ^^ SUPERDUPER, works perfectly. Thanks a lot and have a good weekend!

**Reply by toxsick** [reply]
Hey @hayes,

one more thing regarding the `totalCount` field, by adding the `_count` block in the `select` function:

![pothos 2023-02-16 11-58-25](https://user-images.githubusercontent.com/951466/219346741-db95321f-d44a-4dbd-bc16-2fedec40c023.png)

This will **always** perform a count on the table, even if the `totalCount` field is not requested in the query. In my case this has a pretty significant  performance hit (5ms VS 1976ms).

Can you think of a way to include the `_count` only if the `totalCount` field is requested?

regards

**Reply by hayes** [reply]
You could just move the query into the totalCount resolver. If you aren't filtering or reusing the connection that should work fine. 

The 2 other options are to make totalCount a function on the returned connection object that can load the count (this allows you to access args for the connection) and then call that function in the totalCount resolver.  Or you can use the info arg to look for a selection of the count field to conditionally select the count.  This can get slightly complex, but if you don't need to worry about fragments it's pretty easy
